doc/user/content/sql/create-source/kafka.md

```diff
@@ -0,0 +1,371 @@
+{{</ note >}}
+
+## Syntax
+
+{{< diagram "create-source-kafka.svg" >}}
+
+#### `format_spec`
+
+{{< diagram "format-spec.svg" >}}
+
+#### `key_constraint`
+
+{{< diagram "key-constraint.svg" >}}
+
+#### `with_options`
+
+{{< diagram "with-options.svg" >}}
+
+{{% create-source/syntax-connector-details connector="kafka" envelopes="debezium upsert append-only" %}}
+
+### `WITH` options
+
+Field                                | Value     | Description
+-------------------------------------|-----------|-------------------------------------
+`client_id`                          | `text`    | Use the supplied value as the Kafka client identifier.
+`group_id_prefix`                    | `text`    | Use the specified prefix in the consumer group ID. The resulting `group.id` looks like `<group_id_prefix>materialize-X-Y`, where `X` and `Y` are values that allow multiple concurrent Kafka consumers from the same topic.
+`ignore_source_keys`                 | `boolean` | Default: `false`. If `true`, do not perform optimizations assuming uniqueness of primary keys in schemas.
+`isolation_level`                    | `text`    | Default: `read_committed`. Controls how to read messages that were transactionally written to Kafka. Supported options are `read_committed` to read only committed messages and `read_uncommitted` to read all messages, including those that are part of an open transaction or were aborted.
+`statistics_interval_ms`             | `int`     | `librdkafka` statistics emit interval in `ms`. A value of 0 disables statistics. Statistics can be queried using the `mz_kafka_source_statistics` system table. Accepts values [0, 86400000].
+`timestamp_frequency_ms`             | `int`     | Default: `1000`. Sets the timestamping frequency in `ms`. Reflects how frequently the source advances its timestamp. This measure reflects how stale data in views will be. Lower values result in more-up-to-date views but may reduce throughput.
+`topic_metadata_refresh_interval_ms` | `int`     | Default: `300000`. Sets the frequency in `ms` at which the system checks for new partitions. Accepts values [0,3600000].
+`enable_auto_commit`                 | `boolean` | Default: `false`. Controls whether or not Materialize commits read offsets back into Kafka. This is purely for consumer progress monitoring and does not cause Materialize to resume reading from where it left off across restarts.
+`fetch_message_max_bytes` | `int` | Default: `134217728`. Controls the initial maximum number of bytes per topic+partition to request when fetching messages from the broker. If the client encounters a message larger than this value it will gradually try to increase it until the entire message can be fetched. Accepts values [1, 1000000000].
+
+## Supported formats
+
+|<div style="width:290px">Format</div> | Append-only envelope | Upsert envelope | Debezium envelope |
```

Comment:
Should `Debezium upsert` be a separate column here? It sorta is another envelope.

Reply:
I never realized until now that there were two distinct code paths for `ENVELOPE UPSERT` and `ENVELOPE DEBEZIUM UPSERT`, but it seems like they have been merged recently in #10730. From a user perspective, making the distinction instead of pointing out you can combine both would introduce unnecessary confusion, IMO.

As a side note, `ENVELOPE MATERIALIZE` is also not listed explicitly to avoid confusion (at least until we have a clear idea of where to take that).

Reply:
Hmmm, I can see that, but I think the docs right are confusing in the other direction: `UPSERT` seems like a modifier of `DEBEZIUM`, so it's hard (for me!) to understand if `DEBEZIUM UPSERT` should follow the rules of `UPSERT` or `DEBEZIUM` or a weird mix of the two.

I also don't know enough about the performance implications of `DEBEZIUM UPSERT`, which makes me more willing to treat it as a completely separate thing that happens to contain the name of two other envelope types.

Reply:
Fully agree on `ENVELOPE MATERIALIZE` being out, though: I think we need to make some product decisions on where we're going before publicizing it again.

Reply:
> to treat it as a completely separate thing that happens to contain the name of two other envelope types.

This is my model of `DEBEZIUM UPSERT` as well.

Reply:
Here is the way to think about it.

## What `UPSERT` does

Materialize wants to ingest a **keyed** (requirement for upsert) collection which we internally we represent as diffs. The three basic operations of `insert`, `update`, `delete` are represented as `(key, new_value, +1)`, `(key, new_value, +1), (key, old_value, -1)`, and `(key, old_value, -1)` respectively.

Upstream sources are great at providing the information to generate the `+1` diffs but often cut corners when representing the `-1` diffs or omit it altogether. For example in plain kafka topic the way to represent an insert and an update is the same. You just write the new value in the topic and it is implied that whoever is reading the topic should consider previous values for the same key deleted (if they exist). And the way you represent a delete is a tombstone that looks like `(key, <missing>, -1)`.

`UPSERT` is the process that fills in the missing information by remembering what the current value for each key is. Specifically it has the following logic:
* When it sees a `(value, +1)` for a particular `key` and:
    * we haven't seen that key before it translates it to `(key, value, +1)` (insert)
    * we have seen a value before it translates it to `(key, value, +1), (key, old_value, -1)` (update)
* When it sees a `(value, -1)` for a particular `key` and:
    * we haven't seen that key before it ignores it (no-op delete)
    * we have seen a value before it translates it to `(key, value, -1)` (delete)

## What `DEBEZIUM` does

Debezium is great at providing both positive and negative diffs. Each record it publishes in kafka looks like `(before_value, after_value)` so we can directly map the former to `(key, before, -1)` and the latter to `(key, after, +1)` (an update). It represents inserts and deletes by making the `before` or `after` field `null` respectively. This works great for **uncompacted** topics because it is guaranteed that if you see a `before` value for a key there must have been a previous record in the topic which had the exact same value in the `after` field, so they cancel out.

However, when compaction is enabled we can get into trouble. Imagine a topic with a single key that first gets inserted and then updated, so the data looks like this: `(before=null, after="a"),(before="a", after="b")`. Then, the topic get compacted and contains just `(before="a", after="b")`. If we were to ingest this topic and generate a diff of `("a", -1)` it would be incorrect, as the corresponding positive diff has not been produced! Thankfully, we have some logic that deals with exactly this problem..

## What `DEBEZIUM UPSERT` does

Essentially this is applying BOTH envelopes. First we pass the data into the `DEBEZIUM` envelope logic whose job is to unpack the `before` and `after` fields and turn them into differential dataflow updates. The stream is then passed into the `UPSTREAM` logic that will correct any anomalies (like deleting something that never existed). That is all to say that this is not so much a modifier but literally adding an additional step in the pipeline.

-----

I'm not sure what is the best way to explain this to users but hopefully this clarifies what these envelopes mean to the docs team to think about it. A succint way of saying it is that if your diffs don't match up for whatever reason you need to put an UPSERT at the end to correct them at the expense of more memory.


**PS: To prevent confusion: Our current code doesn't reflect 100% what I wrote above but this description is how it should and how it will work as we progress in the implementation. The differences of the current implementation are minor and not worth elaborating here or in the docs imo.**

Reply:
Thanks for clarifying, @petrosagg! I'll move forward with the current state of things, then. AFAIU, what is currently under `Using Debezium` + the CDC guides is in line with the above (_i.e._ pointing out the cases when you should combine `ENVELOPE DEBEZIUM`+`UPSERT`, but not singling out `ENVELOPE DEBEZIUM UPSERT` a standalone envelope). 

If the mood changes, let's address it in a follow-up revision of the page!